File: 图解算法数据结构/Offer05替换空格/Offer05.py
class Offer05:
    def replaceSpace(self, s: str) -> str:
        res = []#初始化list
        for c in s:#遍历s中的每个字符c
            if c == ' ': 
                res.append("%20")#当字符为空时，往list中添加%20
            else: 
                res.append(c)#当字符不为空时，往list中添加字符c
        return "".join(res)

result=Offer05()
s = "We are happy."
print(result.replaceSpace(s))

File: 图解算法数据结构/Offer05替换空格/Offer05.cpp
#include<iostream>
#include<string>//用来定义字符串
using namespace std;
class Offer05 {
public:
    string replaceSpace(string s) {
        int count = 0;
        int len = s.size();
        // 统计空格数量
        for (char c : s) {
            if (c == ' ') count++;
        }
        // 修改 s 长度,腾出两个字符的空位插入原来的长度中
        s.resize(len + 2 * count);
        // 倒序遍历修改
        for(int i = len - 1, j = s.size() - 1; i < j; i--, j--) {
            if (s[i] != ' ')
                s[j] = s[i];
            else {
                s[j - 2] = '%';
                s[j - 1] = '2';
                s[j] = '0';
                j -= 2;
            }
        }
        return s;
    }
};
int main()
{
    Offer05 result;
    string s;
    s = "We are happy.";
    cout<<result.replaceSpace(s)<<endl;
    return 0;
}
